llvm/lib/Target/DirectX/DXIL.td

```diff
@@ -293,6 +293,43 @@ class Attributes<Version ver = DXIL1_0, list<DXILAttribute> attrs> {
   list<DXILAttribute> op_attrs = attrs;
 }
 
+class DXILConstant<int value_> {
+  int value = value_;
+}
+
+defset list<DXILConstant> BarrierModes = {
```

Comment:
Is `DXILConstant` added out of necessity? Or can you do `defset list<int> BarrierModes ...`?
Was it added for better documentation/extensibility?

Otherwise, we could remove `index` from the `Arg` class and replace the the `is...` bits with a type enum to match the `ArgSelect` struct directly and I think it would allow for more concise code in `DXILEmitter.cpp`

Reply:
It is. I don't think there's a way for definitions to be just a primitive type like int; it has to be instantiation of a class, which in turn has int members. I could be wrong though.